README.md
# TYPO3 - Translation Utils

A simple development helper package which provides some command line utilities to work with translation files.

**A word of caution:** This is a development helper and should not be installed on a production machine!

## Requirements

- TYPO3 v10
- [TYPO3 - Better API](https://github.com/labor-digital/typo3-better-api)
- Installation using Composer

## Installation

Install this package using composer:

```
composer require labor-digital/typo3-translation-utils --dev
```

After that, you can activate the extension in the Extensions-Manager of your TYPO3 installation

## Documentation

#### t3tu:export <EXT_KEY>

Exports the translation labels of an extension into a csv file. A csv file will be created for every base translation file in your extension. Meaning
locallang.xlf will be exported into locallang.csv, while locallang_be.xlf will be exported into locallang_be.csv.

The script will load all translations in your directory, that are named de.locallang.xlf, es.locallang.xlf,... and put the labels into their own column of the
file.

With this you can easily provide your Translators an overview of all translations and their matching key.

##### Option: --format | -f

Allows you to change the output format of the translations. By default a .csv file is exported. With this option you can also set it to .xls, .xlsx or .ods,
depending on your needs

#### t3tu:import <EXT_KEY>

Imports the csv files of a translation into xlf translation files. This is basically the reverse operation to "export". After you changed the content in your
csv files in your translation directory. This command will read all the csv files and update/create the required translation xlf files.

**NOTE**
If the script finds .xls, .xlsx or .ods files in your language directory, it will parse them in the same way it does .csv files. This is helpful if you have
issues with UTF-8 encoded chars in your excel file

#### t3tu:sync <EXT_KEY>

Synchronizes all translation files (e.g. de.locallang.xlf,...) with your origin file (e.g. locallang.xlf) and vice versa.

The sync uses the origin file, as "source of truth", meaning all translation keys you add or remove there will be added/removed to the matching translation
files. All newly created entries in the translation files are prefixed with COPY FROM - $SOURCELANG: to make them easy to spot.

If you rename a translation key in the origin file, the keys in the translation files will be updated automatically to match.
(ONLY if the "source" tag matches both the origin file and the translation file, otherwise the key in the translation file will be recreated!).

The sync will automatically update the "source" tag in your translation files if it was changed in the origin file.

The sync will also make sure that all source files, have the same language variants available to them. This means in practice, if you start with:

- locallang.xlf
- de.locallang.xlf
- locallang_be.xlf
- es.locallang_be.xlf

After the sync you will get something like this:

- locallang.xlf
- de.locallang.xlf
- es.locallang.xlf
- locallang_be.xlf
- de.locallang_be.xlf
- es.locallang_be.xlf

You can also introduce a new language variant by adding an empty file for your language. The file gets automatically prepared with a dummy content and added to
all existing origin files.

### Configuration

You can configure the extension using the "ext config" framework provided by the [TYPO3 - Better API](https://github.com/labor-digital/typo3-better-api)
extension. To create a config, add a new class under `$ext_key/Configuration/ExtConfig/TranslationUtils.php` which looks like this:

```php
<?php
namespace LaborDigital\T3baExample\Configuration\ExtConfig;
use LaborDigital\T3ba\ExtConfig\ExtConfigContext;
use LaborDigital\T3tu\ExtConfigHandler\ConfigureTranslationUtilsInterface;
use LaborDigital\T3tu\ExtConfigHandler\TransUtilsConfigurator;

class TranslationUtils implements ConfigureTranslationUtilsInterface
{
    public static function configureTranslationUtils(TransUtilsConfigurator $configurator, ExtConfigContext $extConfigContext): void
    {
        // The default export format can be configured for the extension
        $configurator->setDefaultExportFormat('xls');
        
        // You can also add constraints that define which languages and/or files should be included by specific actions or not.
        $configurator->exportConstraint()
                     ->addAllowedLanguage('en,es')
                     ->addIgnoredFile('locallang_mod_module.xlf');
        
        // As a rule of thumb, the "allow-list" always beats the "ignore-list". So in this example, we ignore spanish
        // but explicitly allow it for the "locallang.xlf" file when the t3tu:sync command is executed.
        $configurator->syncConstraint()
                     ->addIgnoredLanguage('es')
                     ->addIgnoredFile('locallang_be')
                     ->addAllowedFile('locallang.xlf', 'es');
    }
    
}
```

#### Configuring other extensions

The configuration is, by default, constrainted to the extension which adds the configuration class. So Our `T3baExample` namespace above, is only active for
the `t3ba_example` extension. To provide a configuration for other extensions, you can use the "runWithExtKeyAndVendor"
simulator provided by the ext config implementation. Using this method, you can provide configuration as if added by another extension. So to configure ext:news
you could do this:

```php
<?php

namespace LaborDigital\T3baExample\Configuration\ExtConfig;


use LaborDigital\T3ba\ExtConfig\ExtConfigContext;
use LaborDigital\T3tu\ExtConfigHandler\ConfigureTranslationUtilsInterface;
use LaborDigital\T3tu\ExtConfigHandler\TransUtilsConfigurator;

class TranslationUtils implements ConfigureTranslationUtilsInterface
{
    public static function configureTranslationUtils(TransUtilsConfigurator $configurator, ExtConfigContext $extConfigContext): void
    {
        $extConfigContext->runWithExtKeyAndVendor('news', null, function () use ($configurator) {
            // Everything in this closure is executed as if the configuration was added by ext:news
            $configurator->setDefaultExportFormat('xls');
        });
    }
}

```

## Postcardware

You're free to use this package, but if it makes it to your production environment we highly appreciate you sending us a postcard from your hometown, mentioning
which of our package(s) you are using.

Our address is: LABOR.digital - Fischtorplatz 21 - 55116 Mainz, Germany

We publish all received postcards on our [company website](https://labor.digital).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [10.0.5](https://github.com/labor-digital/typo3-translation-utils/compare/v10.0.4...v10.0.5) (2022-04-29)


### Bug Fixes

* only update file if the content was changed ([6ed3abd](https://github.com/labor-digital/typo3-translation-utils/commit/6ed3abd8bac33e90356334b74cc04efaf4cbbfc1))

### [10.0.4](https://github.com/labor-digital/typo3-translation-utils/compare/v10.0.3...v10.0.4) (2021-12-22)


### Bug Fixes

* **File\Writer:** ensure that xml style does not create tabs ([6225a90](https://github.com/labor-digital/typo3-translation-utils/commit/6225a9061d93c0189c46cc67dc147fc465fe3d99))

### [10.0.3](https://github.com/labor-digital/typo3-translation-utils/compare/v10.0.2...v10.0.3) (2021-08-26)


### Bug Fixes

* resolve multiple issues with the new node format ([9986432](https://github.com/labor-digital/typo3-translation-utils/commit/9986432507114fa475bdefd18ded9c55e4b724cd))

### [10.0.2](https://github.com/labor-digital/typo3-translation-utils/compare/v10.0.1...v10.0.2) (2021-08-25)


### Bug Fixes

* **SyncMapper:** use class to determine node type ([f1b3d28](https://github.com/labor-digital/typo3-translation-utils/commit/f1b3d28c5dac3a5a8dac63a577ebc12935c9243f))
* remove .editorconfig ([20d336d](https://github.com/labor-digital/typo3-translation-utils/commit/20d336d2b6882e2227e862e184b0728b1f311a28))

### [10.0.1](https://github.com/labor-digital/typo3-translation-utils/compare/v10.0.0...v10.0.1) (2021-08-23)

## [10.0.0](https://github.com/labor-digital/typo3-translation-utils/compare/v9.2.1...v10.0.0) (2021-08-23)

### ⚠ BREAKING CHANGES

* the extension now requires TYPO3 v10

### Features

* migrate to TYPO3 v10 ([ccdac76](https://github.com/labor-digital/typo3-translation-utils/commit/ccdac7625cc7b32b030ba856fca8e516cb2fdc7b))

### [9.2.1](https://github.com/labor-digital/typo3-translation-utils/compare/v9.2.0...v9.2.1) (2020-11-02)

### Bug Fixes

* handle labels with <![CDATA content
  correctly ([1872bcc](https://github.com/labor-digital/typo3-translation-utils/commit/1872bcc8d2d782215e04b7055a530a763fa8ad6e))

## [9.2.0](https://github.com/labor-digital/typo3-translation-utils/compare/v9.1.0...v9.2.0) (2020-10-07)

### Features

* implement import/export support for excel and open document
  formats ([4aab991](https://github.com/labor-digital/typo3-translation-utils/commit/4aab99180908cc23d5abdf6bf78b2876de39a922))

### Bug Fixes

* **ImportExport:** remove no longer required write() method from
  TranslationSpreadSheetFile ([8faeffa](https://github.com/labor-digital/typo3-translation-utils/commit/8faeffaa069beba0bbb23cb11625c434d4956e13))

## 9.1.0 (2020-08-06)

### Features

* initial release ([e7cacea](https://github.com/labor-digital/typo3-translation-utils/commit/e7cacea5e58355964efdaa7840fe4b76b38b72cf))


ext_emconf.php
<?php
/**
 * Copyright 2020 LABOR.digital
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Last modified: 2020.07.20 at 22:21
 */

$EM_CONF[$_EXTKEY] = [
    "title" => "LABOR - Typo3 - Translation Utils",
    "description" => "A set of TYPO3 cli tasks for working with translation labels",
    "author" => "Martin Neundorfer",
    "author_email" => "m.neundorfer@labor.digital",
    "category" => "search",
    "author_company" => "Labor.digital",
    "shy" => "",
    "conflicts" => "",
    "priority" => "",
    "module" => "",
    "state" => "stable",
    "internal" => "",
    "uploadfolder" => 0,
    "createDirs" => "",
    "modify_tables" => "",
    "clearCacheOnLoad" => 1,
    "lockType" => "",
    "version" => "10.0.5",
    "constraints" => [
        "depends" => [
            "typo3" => "10.0.0-10.99.99",
        ],
        "conflicts" => [
        ],
        "suggests" => [
        ],
    ],
    "suggests" => [
    
    ],
];
